### ISSUE TITLE
User supplied flexible checksums not respected

### ISSUE THREAD

**Issue by aajtodd** [awslabs/aws-sdk-rust#1122]
### Describe the bug

When a user supplies both `checksum_algorithm` and a pre-calculated checksum the SDK will go ahead and calculate one anyway and overwrite the user supplied one.

```rust
s3.put_object()
    .bucket(bucket_name)
    .key(key_name)
    .checksum_algorithm(aws_sdk_s3::types::ChecksumAlgorithm::Sha256)
    .checksum_sha256("bad-sha")
    .body(body)
    .send()
    .await?;
```

### Expected Behavior

The SDK should detect this and use the user supplied value and do nothing. This is the behavior described in the SEP.

### Current Behavior

The user supplied checksum is serialized to a header and then is later overwritten with the value it calculates.

### Reproduction Steps

See bug description. Sending a bad value for the checksum should fail but it doesn't because we overwrite it.

### Version

```text
aws-sdk-s3 = "1.21.0"
```

### Environment details (OS name and version, etc.)

macOS 14.4

**Comment by CNorskov** [comment]
Same problem here (with a working solution):
https://github.com/awslabs/aws-sdk-rust/issues/1103#issuecomment-2008959603

In short:
A workaround is not specifying the algorithm. So a working code would be:

```rust
s3.put_object()
    .bucket(bucket_name)
    .key(key_name)
    .checksum_sha256("bad-sha")
    .body(body)
    .send()
    .await?;
```

**Comment by aajtodd** [comment]
closing as duplicate of #1103
